scheduler/localsync.go

```diff
@@ -10,83 +10,87 @@
-	zkservice "github.com/control-center/serviced/zzk/service"
-	zkvirtualips "github.com/control-center/serviced/zzk/virtualips"
+	"github.com/control-center/serviced/domain/host"
+	"github.com/control-center/serviced/domain/pool"
+	"github.com/control-center/serviced/domain/service"
 	"github.com/zenoss/glog"
 )
 
-const (
-	minWait = 30 * time.Second
-	maxWait = 3 * time.Hour
-)
-
-func (s *scheduler) localSync(shutdown <-chan interface{}, rootConn client.Connection) {
-	wait := time.After(0)
-
-retry:
-	for {
-		select {
-		case <-wait:
-		case <-shutdown:
-			return
-		}
+// LocalSyncDatastore contains the primary datastore information from which to
+// sync.
+type LocalSyncDatastore interface {
+	// GetResourcePools returns all the resource pools
+	GetResourcePools() ([]pool.ResourcePool, error)
+	// GetHosts returns hosts for a particular resource pool
+	GetHosts(poolID string) ([]host.Host, error)
+	// GetServices returns services for a particular resource pool
+	GetServices(poolID string) ([]service.Service, error)
+}
 
-		pools, err := s.GetResourcePools()
-		if err != nil {
-			glog.Errorf("Could not get resource pools: %s", err)
-			wait = time.After(minWait)
-			continue
-		} else if err := zkservice.SyncResourcePools(rootConn, pools); err != nil {
-			glog.Errorf("Could not do a local sync of resource pools: %s", err)
-			wait = time.After(minWait)
-			continue
-		}
+// LocalSyncInterface is the endpoint where data is synced
+type LocalSyncInterface interface {
+	// SyncResourcePools synchronizes resource pools
+	SyncResourcePools(pools []pool.ResourcePool) error
+	// SyncVirtualIPs synchronizes virtual ips for a pool
+	SyncVirtualIPs(poolID string, virtualIPs []pool.VirtualIP) error
+	// SyncHosts synchronizes hosts for a pool
+	SyncHosts(poolID string, hosts []host.Host) error
+	// SyncServices synchronizes services for a resource pool
+	SyncServices(poolID string, svcs []service.Service) error
+}
 
-		for _, pool := range pools {
-			conn, err := zzk.GetLocalConnection(zzk.GeneratePoolPath(pool.ID))
-			if err != nil {
-				glog.Errorf("Could not get a pool-based connection for %s to zookeeper: %s", pool.ID, err)
-				wait = time.After(minWait)
-				continue retry
-			}
+// LocalSync performs synchronization from the primary datastore to the
+// coordinator
+type LocalSync struct {
+	ds    LocalSyncDatastore
+	iface LocalSyncInterface
+}
 
-			// Update the hosts
-			if hosts, err := s.GetHostsByPool(pool.ID); err != nil {
-				glog.Errorf("Could not get hosts in pool %s: %s", pool.ID, err)
-				wait = time.After(minWait)
-				continue retry
-			} else if err := zkservice.SyncHosts(conn, hosts); err != nil {
-				glog.Errorf("Could not do a local sync of hosts: %s", err)
-				wait = time.After(minWait)
-				continue retry
-			}
+// Purge performs the synchronization for services, hosts, pools, and virtual
+// ip.
+// Implements utils.TTL
+func (sync *LocalSync) Purge(age time.Duration) (time.Duration, error) {
```

Comment:
Are SyncVirtualIPs, SyncHosts, and SyncServices idempotent? i.e., if something goes wrong in the loop starting on line 69 and this function exits prematurely, is it going to do something screwy to the data when it runs again?